File: android/src/main/java/org/ligi/passandroid/model/comparator/PassByTimeComparator.java
package org.ligi.passandroid.model.comparator;

import java.util.Comparator;
import org.ligi.passandroid.model.pass.Pass;

public class PassByTimeComparator implements Comparator<Pass> {

    final int NO_NULL_VALUE = 5;

    @Override
    public int compare(Pass lhs, Pass rhs) {
        final int nullCheck = calculateCompareForNullValues(lhs, rhs);
        if (nullCheck == NO_NULL_VALUE) {
            return rhs.getCalendarTimespan().getFrom().compareTo(lhs.getCalendarTimespan().getFrom());
        } else {
            return nullCheck;
        }
    }

    public int calculateCompareForNullValues(final Pass lhs, final Pass rhs) {
        if ((lhs.getCalendarTimespan() == null || lhs.getCalendarTimespan().getFrom() == null) &&
            (rhs.getCalendarTimespan() == null || rhs.getCalendarTimespan().getFrom() == null)) {
            return 0;
        }

        if (lhs.getCalendarTimespan() == null || lhs.getCalendarTimespan().getFrom() == null) {
            return 1;
        }
        if (rhs.getCalendarTimespan() == null || rhs.getCalendarTimespan().getFrom() == null) {
            return -1;
        }
        return NO_NULL_VALUE;
    }
}

File: android/src/main/java/org/ligi/passandroid/model/comparator/PassByTimeComparator.kt
package org.ligi.passandroid.model.comparator

import org.ligi.passandroid.model.pass.Pass
import org.threeten.bp.ZonedDateTime
import java.util.*

open class PassByTimeComparator : Comparator<Pass> {

    override fun compare(lhs: Pass, rhs: Pass): Int {
        return calculateCompareForNullValues(lhs, rhs, { leftDate: ZonedDateTime, rightDate: ZonedDateTime ->
            return@calculateCompareForNullValues leftDate.compareTo(rightDate)
        })
    }

    protected fun calculateCompareForNullValues(lhs: Pass, rhs: Pass, foo: (leftDate: ZonedDateTime, rightDate: ZonedDateTime) -> Int): Int {
        val leftDate = extractPassDate(lhs)
        val rightDate = extractPassDate(rhs)

        if (leftDate == rightDate) {
            return 0
        }

        if (leftDate == null) {
            return 1
        }
        if (rightDate == null) {
            return -1
        }
        return foo(leftDate, rightDate)
    }

    private fun extractPassDate(pass: Pass): ZonedDateTime? {
        if (pass.calendarTimespan != null && pass.calendarTimespan!!.from != null) {
            return pass.calendarTimespan!!.from
        }

        if (pass.validTimespans != null && !pass.validTimespans!!.isEmpty()) {
            return pass.validTimespans!![0].from
        }

        return null
    }
}
